cmd/snap-bootstrap: Listen to keyboard added after start and handle switch root

valentindavid opened the pull request:
Keyboard drivers and the drivers for the bus they are plugged into are
not necessarily initialized by the time udev has settled cold plug
events. That means we have to read events for new input devices and
read those keyboards.

Also because the daemon is started in initramfs and needs to write
into the file system, it needs to chroot when switch root happens.

There are 2 things to decide in this PR:
 - What should be the proper way to notify the switch root. This PR uses SIGUSR1. But we could use an RPC for example using dbus. That way we could pass the path too the new root for example.
 - Now we always wait 10s before starting snapd. Before, it would not wait if no keyboard was found. Should we have a second timeout, like if after 3s no keyboard was found, we cancel.
 - `osutil/udev` has another interface to look for device through `/sys`. Should we replace use of `evdev.ListInputDevices` to use that?

I will open a PR on core-initrd later for necessary changes and link to it.

valentindavid commented:
https://github.com/snapcore/core-initrd/pull/97

Meulengracht commented on cmd/snap-bootstrap/triggerwatch/evdev.go:
```diff
@@ -205,6 +205,22 @@ func (e *evdevKeyboardInputDevice) Close() {
+	if !ok {
+		return nil, fmt.Errorf("cannot find a key matching the filter %q", filter.Key)
+	}
+	cap := evdev.CapabilityCode{Code: kc, Name: filter.Key}
```
is it neccessary to declare cap here? it seems like we are only using kc in line 220 - couldn't we just instead do uint16(kc)? Sorry if it is a stupid question and there is something I'm just not seeing

valentindavid replied:
Correct. I only extracted the code needed from FindMatchingDevices. But I think I do not properly do the filtering. I will rework that part.

valentindavid replied:
I have changed the code to something that makes a bit more sense.

alfonsosanchezbeato commented:
Tbh I do not feel very comfortable with any mandatory delay before snapd starts. I think we should have a way to fully disable the recovery chooser trigger to avoid that and maybe even make that the default. @mvo5 wdyt?

Re: using RPC instead of signals, plymouth uses a UNIX socket, maybe that would be the easiest route if we really need that, which I don't think so.

mardy commented on cmd/snap-bootstrap/cmd_recovery_chooser_trigger.go:
```diff
@@ -74,10 +77,19 @@ func (c *cmdRecoveryChooserTrigger) Execute(args []string) error {
 		markerFile = c.MarkerFile
 	}
 	logger.Noticef("trigger wait timeout %v", timeout)
+	logger.Noticef("default timeout %v", deviceTimeout)
```
s/default/device/

valentindavid replied:
Fixed

mardy commented on cmd/snap-bootstrap/triggerwatch/evdev.go:
```diff
@@ -205,6 +205,27 @@ func (e *evdevKeyboardInputDevice) Close() {
+	if err != nil {
+		return nil, err
+	}
+	kc, ok := strToKey[filter.Key]
```
nitpick: please pick a longer name for `kc`; I guess it stands for "key code", but I could also be wrong :-)

mardy commented on cmd/snap-bootstrap/triggerwatch/triggerwatch.go:
```diff
@@ -71,17 +112,40 @@ func Wait(timeout time.Duration) error {
+				return ErrNoMatchingInputDevices
+			}
+		case uevent := <-ueventQueue:
+			dev, err := trigger.Open(triggerFilter, uevent.Env["DEVNAME"])
+			if dev != nil && err == nil {
+				foundDevice = true
+				defer dev.Close()
+				go dev.WaitForTrigger(detectKeyCh)
+			}
```
What happens if the file cannot be opened? Don't we need to return an error, or maybe just to add some line to the log?

valentindavid replied:
We should ignore device that cannot be open. Added a notice in the log.

mardy commented on cmd/snap-bootstrap/triggerwatch/triggerwatch.go:
```diff
@@ -71,17 +112,40 @@ func Wait(timeout time.Duration) error {
+			logger.Noticef("Switching root")
+			err := syscall.Chroot("/sysroot")
+			if err != nil {
+				return errors.New("Cannot change root")
```
I'd also include the original error to this line, maybe it can contain some important hints.

mardy reviewed (approved):
LGTM, thanks! Just minor nitpicks and questions.

alfonsosanchezbeato commented:
> Tbh I do not feel very comfortable with any mandatory delay before snapd starts. I think we should have a way to fully disable the recovery chooser trigger to avoid that and maybe even make that the default. @mvo5 wdyt?
> 

I am also thinking, why do we prevent snapd from starting while we wait for keyboard input? Maybe we could stop the snapd service if the user finally presses a key, if necessary.

Meulengracht reviewed (approved):
Thanks for the changes, LGTM. I really like alfonso's suggestion about maybe starting snapd in parallel, but maybe that is out of scope for this PR?

valentindavid commented:
> I am also thinking, why do we prevent snapd from starting while we wait for keyboard input? Maybe we could stop the snapd service if the user finally presses a key, if necessary.

I can try to make only console-conf depend on it. See if that works.

valentindavid commented:
@alfonsosanchezbeato  I have tested changing the boot dependencies. And it works. Here are the changes.

 - Removed `Before=snapd.service`, from `snapd.recovery-chooser-trigger.service`
 - Added `After=snapd.recovery-chooser-trigger.service` in `getty@.service` and `serial-getty@.service`.

Unfortunately this requires changes in both snapd and all versions of core. So it is a bit unpractical. The main issue is we cannot remove `Before=` from units through drop-ins. So we will need to add `After=...` in `core`, `core18`, `core20`, `core22` before we can add the `Before=...` in snapd.

Meulengracht commented:
> I have found that adding `Wants=getty-pre.target` and `Before=getty-pre.target` works.

What kind of impacts on boot time does this have?

valentindavid commented:
> > I have found that adding `Wants=getty-pre.target` and `Before=getty-pre.target` works.
> 
> What kind of impacts on boot time does this have?

getty is only for the consoles. So that means it removes the impact on everything not depending on the consoles. Which means pretty much everything else than console-conf. As it is now, if there is a keyboard that is detected, it will take 10s before snapd is started. But with this change, it will start as soon as possible.

valentindavid commented on cmd/snap-bootstrap/triggerwatch/triggerwatch.go:
```diff
@@ -71,17 +112,42 @@ func Wait(timeout time.Duration) error {
+			}
+		case <-sigs:
+			logger.Noticef("Switching root")
+			err := syscall.Chroot("/sysroot")
```
Right, I can add chdir. But it worked without it. We only open one file: the tag file.

valentindavid replied:
Fixed.

alfonsosanchezbeato reviewed (approved):
LGTM, thanks for the change

bboozzoo reviewed (approved):
LGTM

valentindavid commented on cmd/snap-bootstrap/triggerwatch/triggerwatch.go:
```diff
@@ -51,7 +65,33 @@ var (
+				Env: map[string]string{
+					"SUBSYSTEM":         "input",
+					"ID_INPUT_KEYBOARD": "1",
+					"DEVNAME":           ".*",
```
An input device has multiple "events" sub devices. Only those "events" devices have `/dev` node. When we plug a keyboard, or mouse, we get an uevent event for the device, then some uevent events for each input "events" nodes (note: not the same kind of event). We want only the ones we can actually read, which is the ones with a `/dev` node. This is why we need to filter uevents to have `DEVNAME` defined. Though the value does not matter.
